tests/miral/test_window_manager_tools.cpp

```diff
@@ -0,0 +1,290 @@
+    void handle_request_resize(miral::WindowInfo& /*window_info*/, MirInputEvent const* /*input_event*/, MirResizeEdge /*edge*/) {}
+    mir::geometry::Rectangle confirm_placement_on_display(const miral::WindowInfo&, MirWindowState, mir::geometry::Rectangle const& new_placement)
+    {
+        return new_placement;
+    }
+};
+
+struct TestDisplayConfigurationObserver : mir::ObserverRegistrar<mir::graphics::DisplayConfigurationObserver>
+{
+    void register_interest(std::weak_ptr<mir::graphics::DisplayConfigurationObserver> const& o) override
+    {
+        ASSERT_THAT(observer.lock(), testing::IsNull())
+            << "TestDisplayConfigurationObserver does not support multiple observers";
+        observer = o;
+    }
+
+    void register_interest(std::weak_ptr<mir::graphics::DisplayConfigurationObserver> const& o, mir::Executor&) override
+    {
+        register_interest(o);
+    }
+
+    void unregister_interest(mir::graphics::DisplayConfigurationObserver const& o) override
+    {
+        ASSERT_THAT(observer.lock().get(), testing::Eq(&o));
+        observer = std::weak_ptr<mir::graphics::DisplayConfigurationObserver>(); // set to null
+    }
+
+    std::weak_ptr<mir::graphics::DisplayConfigurationObserver> observer;
+};
+
+}
+
+namespace mt = mir::test;
+
+struct mt::TestWindowManagerTools::Self
+{
+    StubFocusController focus_controller;
+    StubDisplayLayout display_layout;
+    StubPersistentSurfaceStore persistent_surface_store;
+    TestDisplayConfigurationObserver display_configuration_observer;
+
+};
+
+mt::TestWindowManagerTools::TestWindowManagerTools()
+    : self{std::make_unique<Self>()},
+      session{std::make_shared<StubStubSession>()},
+      window_manager_policy{nullptr},
+      window_manager_tools{nullptr},
+      basic_window_manager{
+        &self->focus_controller,
+        mir::test::fake_shared(self->display_layout),
+        mir::test::fake_shared(self->persistent_surface_store),
+        self->display_configuration_observer,
+        [this](miral::WindowManagerTools const& tools) -> std::unique_ptr<miral::WindowManagementPolicy>
+            {
+                auto policy = std::make_unique<testing::NiceMock<MockWindowManagerPolicy>>(tools);
+                window_manager_policy = policy.get();
+                window_manager_tools = tools;
+                return policy;
+            }
+    }
+{
+}
+
+mt::TestWindowManagerTools::~TestWindowManagerTools() = default;
+
+auto mt::TestWindowManagerTools::create_surface(
+    std::shared_ptr<mir::scene::Session> const& session,
+    mir::scene::SurfaceCreationParameters const& params) -> mir::frontend::SurfaceId
+{
+    // This type is Mir-internal, I hope we don't need to create it here
+    std::shared_ptr<mir::frontend::EventSink> const sink;
+    return session->create_surface(params, sink);
+}
+
+void mt::TestWindowManagerTools::set_outputs(std::vector<miral::Rectangle> outputs)
```

Comment:
Why is set_outputs() doing so much? There seem to be three distinct elements:
1. Creating a DisplayConfiguration double (it happens to be a Mock, but I'm not convinced that is always right);
2. Checking the state of display_configuration_observer; and
3. Sending a configuration_applied notification.

I think it would be more flexible to have:
1. create_mock_display_configuration(outputs) -> shared<DisplayConfiguration>();
2. notify_configuration_applied(shared<DisplayConfiguration>()); and, 
3. possibly now or in the future, create_fake_display_configuration(outputs) -> shared<DisplayConfiguration>() (which does not check that for_each_output() is called exactly once).

(And notify_configuration_applied() can call a notify_configuration_applied() on FakeDisplayConfigurationObserverRegistrar which can worry about its own state?)

Reply:
Done as you suggest. I also changed the expected call to `display_config->for_each_output()` from `WillOnce` to `WillRepeatedly`, as that happening exactly once (or ever) isn't a thing we need to test for.